src/spesh/facts.c

```diff
@@ -606,6 +606,11 @@ static void add_bb_facts(MVMThreadContext *tc, MVMSpeshGraph *g, MVMSpeshBB *bb,
             g->facts[result.reg.orig][result.reg.i].flags |= MVM_SPESH_FACT_DECONTED;
             break;
         }
+        case MVM_OP_nativecallinvokejit: {
+            for (i = 0; i < g->num_locals; i++)
+                g->facts[i][g->fact_counts[i] - 1].usages++;
```

Comment:
This looks weird and fragile. It seems to be giving every local a bump of 1...why? Usage counts are computed up at line 376 for every op.

Reply:
The nativecallinvokejit OP is a bit special as it expects the arguments to be in the first locals of the call frame. Spesh didn't know this and removed them as dead weight. Upping the usage count is done to reflect the actual usage.

Reply:
Ouch, I think what I originally suggested (though it probably wasn't clear why, and so probably sounded odd...) was that `nativecallinvokejit` used the standard args buffer and then was just a kind of `invoke` instruction (so marked `-a` in `oplist` to terminate an args sequence). Then args would be placed into the args buffer like a normal call, and later we'd pull the values out of the `frame->args` buffer instead of `frame-work` (so I suspect it's a relatively small change on that side of things). This way, there'd be no issues with usage counts and so forth, since args in the arg buffer are obviously used by the callee.

Having spent so long of late removing cheats like this in spesh (many driven by the fact that each of them bit back somewhere and cost a bunch of debugging time), I'm quite reluctant to see another one like this go in. Especially as I suspect it's actually not sufficient in the case that a frame with `nativecallinvokejit` in it were to be inlined. If that happens, then we don't re-run the whole facts process on the inlinee, just do some much cheaper usage count setting, and so on that code path we wouldn't do extra usage bumps like is being done here, and then dead code elimination would eat them. We could of course do similar compensation as we do here, but that only makes me fear where we get caught out next...